apiserver/embeddedcommands.go

```diff
@@ -0,0 +1,231 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package apiserver
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/juju/cmd"
+	"github.com/juju/errors"
+	"github.com/juju/names/v4"
+
+	apiservererrors "github.com/juju/juju/apiserver/errors"
+	"github.com/juju/juju/apiserver/params"
+	"github.com/juju/juju/core/model"
+	"github.com/juju/juju/jujuclient"
+)
+
+// ExecEmbeddedCommandFunc defines a function which runs a named Juju command
+// with the whitelisted sub commands.
+type ExecEmbeddedCommandFunc func(ctx *cmd.Context, store jujuclient.ClientStore, whitelist []string, cmd string, args []string) int
+
+// allowedEmbeddedCommands is a whitelist of Juju CLI commands which
+// are permissible to run embedded on a controller.
+var allowedEmbeddedCommands = []string{
+	"actions",
+	"add-machine",
+	"add-relation",
+	"add-space",
+	"add-storage",
+	"add-subnet",
+	"add-unit",
+	"add-user",
+	"agreements",
+	"attach",
+	"attach-resource",
+	"attach-storage",
+	"bind",
+	"cached-images",
+	"cancel-task",
+	"change-user-password",
+	"charm-resources",
+	"clouds",
+	"config",
+	"consume",
+	"controller-config",
+	"create-storage-pool",
+	"credentials",
+	"deploy",
+	"detach-storage",
+	"disable-user",
+	"enable-user",
+	"expose",
+	"find-offers",
+	"firewall-rules",
+	"get-constraints",
+	"get-model-constraints",
+	// TODO(wallyworld) - check if these should be allowed
+	//"grant",
+	//"grant-cloud",
+	"help",
+	"import-filesystem",
+	"machines",
+	"metrics",
+	"model-config",
+	"model-default",
+	"model-defaults",
+	"move-to-space",
+	"offer",
+	"offers",
+	"payloads",
+	"plans",
+	"register",
+	"relate",
+	"reload-spaces",
+	"remove-application",
+	"remove-cached-images",
+	"remove-consumed-application",
+	"remove-credential",
+	"remove-machine",
+	"remove-offer",
+	"remove-relation",
+	"remove-saas",
+	"remove-space",
+	"remove-storage",
+	"remove-storage-pool",
+	"remove-unit",
+	"remove-user",
+	"rename-space",
+	"resolved",
+	"resolve",
+	"resources",
+	"resume-relation",
+	"retry-provisioning",
+	"revoke",
+	"run",
```

Comment:
This is run action rather than exec?

Reply:
Yeah, "run" is "run-action" in new CLI, and equivalent to "exec" in the old CLI. So it exists either way.